packages/block-library/src/navigation/placeholder.js

```diff
@@ -78,28 +78,51 @@ function getSelectedMenu( selectedCreateOption ) {
 /**
  * A recursive function that maps menu item nodes to blocks.
  *
- * @param {Object[]} nodes An array of menu items.
- *
+ * @param {Object[]} menuItems An array of menu items.
  * @return {WPBlock[]} An array of blocks.
  */
-function mapMenuItemsToBlocks( nodes ) {
-	return nodes.map( ( { title, type, link: url, id, children } ) => {
-		const innerBlocks =
-			children && children.length ? mapMenuItemsToBlocks( children ) : [];
+function mapMenuItemsToBlocks( menuItems ) {
+	return menuItems.map( ( menuItem ) => {
+		if ( menuItem.type === 'block' ) {
+			const [ block ] = parse( menuItem.content.raw );
 
-		return createBlock(
-			'core/navigation-link',
-			{
-				type,
-				id,
-				url,
-				label: ! title.rendered
-					? __( '(no title)' )
-					: escape( title.rendered ),
-				opensInNewTab: false,
-			},
-			innerBlocks
-		);
+			if ( ! block ) {
+				return createBlock( 'core/freeform', {
+					content: menuItem.content,
+				} );
+			}
+
+			return block;
+		}
+
+		const attributes = {
+			label: ! menuItem.title.rendered
+				? __( '(no title)' )
+				: escape( menuItem.title.rendered ),
+			opensInNewTab: menuItem.target === '_blank',
+		};
+
+		if ( menuItem.url ) {
+			attributes.url = menuItem.url;
+		}
+
+		if ( menuItem.description ) {
+			attributes.description = menuItem.description;
+		}
+
+		if ( menuItem.xfn?.length && some( menuItem.xfn ) ) {
+			attributes.rel = menuItem.xfn.join( ' ' );
+		}
+
+		if ( menuItem.classes?.length && some( menuItem.classes ) ) {
+			attributes.className = menuItem.classes.join( ' ' );
+		}
+
```

Comment:
Ideally we'd also set `id` and `type` here so that the URL can be dynamic. Need to think about these attributes some more, though. I've added a note to do this in 

I'll add a note to come back to this in https://github.com/WordPress/gutenberg/issues/18345.

Reply:
I'm unclear on the reasoning behind dropping these attributes, what's the background?

Reply:
So the existing code in `master` has a few problems:

- We're setting the `type` attribute to be `menuItem.type`, but this is incorrect because the attribute should be one of `'page'`, `'post'`, `'category'`, `'tag'`, or `undefined` and `menuItem.type` returns `'post_object'` or `'taxonomy'` IIRC.

- We're setting `id` to `menuItem.id`, but this is incorrect because for a link to dynamically render its URL it will need the _ID of the post_, not the _ID of the menu item_. It should probably be set to `menuItem.object_id`.

Instead of fixing these issues, though, I added a note to address them when we address https://github.com/WordPress/gutenberg/issues/18345 as there's some thinking to be done around what attributes are needed for dynamically rendering a link. IMO we should be storing three things with the block: the menu item type, the object type, the object ID. See https://github.com/WordPress/gutenberg/issues/18345#issuecomment-682335969.

Reply:
Thanks for the explanation, sounds like it's not something to tackle here.

It'd be good to rewrite the description on the issue with these details. 👍